File: Oc-Dome/Oc-Dome/Classes/Home(首页)/ChildController(子控制器)/Sorket服务器端/Controller/GCDSocketServerViewController.m
//
//  GCDSocketServerViewController.m
//  Oc-Dome
//
//  Created by 李含文 on 2019/5/25.
//  Copyright © 2019年 东莞市三心网络科技有限公司. All rights reserved.
//

#import "GCDSocketServerViewController.h"
#import "GCDAsyncSocket.h"
#import "PublicTool.h"

#import <ifaddrs.h>
#import <arpa/inet.h>
#import <SystemConfiguration/CaptiveNetwork.h>
#import <CoreTelephony/CTCellularData.h>

@interface GCDSocketServerViewController ()<GCDAsyncSocketDelegate>
@property (weak, nonatomic) IBOutlet UITextField *portTF;
@property (weak, nonatomic) IBOutlet UITextField *msgTF;
@property (weak, nonatomic) IBOutlet UITextView *logTV;

@property (weak, nonatomic) IBOutlet UILabel *ipLabel;


@property (nonatomic, strong)GCDAsyncSocket *serverSocket;
@property (nonatomic, strong)GCDAsyncSocket *clientSocket;


@property (nonatomic, strong)NSMutableArray *clientArr;//连接池
@property (nonatomic, strong)NSMutableArray *clientInfoArr;//客户端信息

@end

@implementation GCDSocketServerViewController

- (NSMutableArray*)clientArr{
    if (_clientArr == nil) {
        _clientArr = [NSMutableArray array];
    }
    return _clientArr;
}

- (NSMutableArray*)clientInfoArr{
    if (_clientInfoArr == nil) {
        _clientInfoArr = [NSMutableArray array];
    }
    return _clientInfoArr;
}
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    
}
- (void)viewDidLoad {
    [super viewDidLoad];
    UITapGestureRecognizer* tap = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(tapToResign:)];
    [self.view addGestureRecognizer:tap];
    [self setupServerSocket];
    NSString *wifiName = [self getCurreWiFiSsid];
    NSString *ip = [self getCurrentLocalIP];
    self.ipLabel.text = [NSString stringWithFormat:@"WIFI:%@ IP:%@", wifiName, ip];
}

- (void)tapToResign:(UITapGestureRecognizer*)tap{
    [_msgTF resignFirstResponder];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)setupServerSocket{
    //在主线程里面回调
    self.serverSocket = [[GCDAsyncSocket alloc]initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
}

- (void)showLogMsg:(NSString*)log{
    _logTV.text = [_logTV.text stringByAppendingFormat:@"%@\n",log];
}


#pragma mark - Event Handle
- (IBAction)btnListenClick:(id)sender {//监听
    NSError* error = nil;
    [self.serverSocket acceptOnPort:_portTF.text.intValue error:&error];
    if (error != nil) {
        NSLog(@"监听出错：%@",error);
    }
    else{
        [self showLogMsg:@"正在监听..."];
    }
}

- (IBAction)btnSendClick:(id)sender {
    [self showLogMsg:[NSString stringWithFormat:@"你: %@",_msgTF.text]];
    NSData* data = [_msgTF.text dataUsingEncoding:NSUTF8StringEncoding];
    //给对应的客户端发送数据
    [self.clientSocket writeData:data withTimeout:-1 tag:0];
    _msgTF.text = @"";
}


#pragma mark - GCDAsyncSocketDelegate

//接收到请求
- (void)socket:(GCDAsyncSocket *)sock didAcceptNewSocket:(GCDAsyncSocket *)newSocket{
    [self showLogMsg:@"收到客户端连接...."];
    NSString *ip = newSocket.connectedHost;
    int port = newSocket.connectedPort;
    [self showLogMsg:[NSString stringWithFormat:@"客户端地址：%@,客户端端口：%d",ip,port]];
    
    
    //收到连接，保存连接到连接池
    NSMutableDictionary* dicClient = [NSMutableDictionary dictionary];
    [dicClient setValue:newSocket forKey:@"socket"];
    [dicClient setValue:newSocket.connectedHost forKey:@"host"];
    
    //排重
    int tempI = -1;
    for (int i=0; i<self.clientArr.count; i++) {
        NSDictionary* client = [self.clientArr objectAtIndex:i];
        if ([client[@"host"] isEqualToString:newSocket.connectedHost]) {
            tempI = i;
        }
    }
    if (tempI >= 0) {
        [self.clientArr removeObjectAtIndex:tempI];
    }
    [self.clientArr addObject:dicClient];
    self.clientSocket = newSocket;
    [newSocket readDataWithTimeout:-1 tag:0];
    [self showLogMsg:[NSString stringWithFormat:@"你: %@",@"你连接成功了,大兄弟"]];
    NSData* data = [@"你连接成功了,大兄弟" dataUsingEncoding:NSUTF8StringEncoding];
    //给对应的客户端发送数据
    [self.clientSocket writeData:data withTimeout:-1 tag:0];
}

//读取信息
- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag{
    NSString* strMsg = [[NSString alloc]initWithData:data encoding:NSUTF8StringEncoding];
    [self showLogMsg:[NSString stringWithFormat:@"客户端:%@",strMsg]];
    [sock readDataWithTimeout:-1 tag:0];
}

- (nullable NSString*)getCurrentLocalIP {
    NSString *address = nil;
    struct ifaddrs *interfaces = NULL;
    struct ifaddrs *temp_addr = NULL;
    int success = 0;
    // retrieve the current interfaces - returns 0 on success
    success = getifaddrs(&interfaces);
    if (success == 0) {
        // Loop through linked list of interfaces
        temp_addr = interfaces;
        while(temp_addr != NULL) {
            if(temp_addr->ifa_addr->sa_family == AF_INET) {
                // Check if interface is en0 which is the wifi connection on the iPhone
                if([[NSString stringWithUTF8String:temp_addr->ifa_name] isEqualToString:@"en0"]) {
                    // Get NSString from C String
                    address = [NSString stringWithUTF8String:inet_ntoa(((struct sockaddr_in *)temp_addr->ifa_addr)->sin_addr)];
                }
            }
            temp_addr = temp_addr->ifa_next;
        }
    }
    // Free memory
    freeifaddrs(interfaces);
    return address;
}
- (nullable NSString *)getCurreWiFiSsid {
    NSArray *ifs = (__bridge_transfer id)CNCopySupportedInterfaces();
    NSLog(@"Supported interfaces: %@", ifs);
    id info = nil;
    for (NSString *ifnam in ifs) {
        info = (__bridge_transfer id)CNCopyCurrentNetworkInfo((__bridge CFStringRef)ifnam);
        NSLog(@"%@ => %@", ifnam, info);
        if (info && [info count]) { break; }
    }
    return [(NSDictionary*)info objectForKey:@"SSID"];
}
@end

File: Swift-Dome/LHWDome/Classes/Example(案例)/ChlidController(子控制器)/Sorket服务器端/Controller/GCDSocketServerViewController.swift
//
//  GCDSocketServerViewController.swift
//  LHWDome
//
//  Created by 李含文 on 2019/5/25.
//  Copyright © 2019年 李含文. All rights reserved.
//

import UIKit
import SystemConfiguration


class GCDSocketServerViewController: BaseViewController {
    @IBOutlet weak var portTF: UITextField!
    @IBOutlet weak var msgTF: UITextField!
    @IBOutlet weak var logTV: UITextView!
    @IBOutlet weak var ipLabel: UILabel!
    /// 服务器
    private var serverSocket : GCDAsyncSocket?
    /// 客户端
    private var clientSocket : GCDAsyncSocket?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        HWPrint("路由器IP:\(HWWIFITools.hw_getRouterIP())")
        let ip = HWWIFITools.hw_getCurrentIP() ?? ""
        let name = HWWIFITools.hw_getCurreWiFiSsid() ?? ""
        ipLabel.text = "WIFI名:\(name) ip:\(ip)"
        HWPrint(name + ip)
        let hotSpotsip = HWWIFITools.hw_getHotSpotsType() ?? ""
        if hotSpotsip.count == 0 {
            HWPrint("热点关闭")
        } else {
            let name = UIDevice.current.name //设备名称
            HWPrint("热点开启")
            ipLabel.text = "WIFI名:\(name) ip:\(hotSpotsip)"
        }
        
        setupServerSocket()
    }
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }
    // MARK: - 创建服务器Socket
    func setupServerSocket() {
        self.serverSocket = GCDAsyncSocket(delegate: self, delegateQueue: DispatchQueue.main)
    }
    
    func showLogMsg(_ log:String) {
        logTV.text = logTV.text + "\n" + log
    }
    // MARK: - 监听按钮点击
    @IBAction func listeningBtnClick(_ sender: Any) {
        do {
            try self.serverSocket?.accept(onPort: UInt16(self.portTF.text ?? "")!)
        }
        catch {
            HWPrint("监听出错")
            return
        }
        showLogMsg("正在监听...")
    }
    
    @IBAction func sendBtnClick(_ sender: Any) {
        if let data = msgTF.text?.data(using: String.Encoding.utf8) {
            showLogMsg(String(format: "你: %@", msgTF.text ?? ""))
            self.clientSocket?.write(data, withTimeout: -1, tag: 0)
            msgTF.text = ""
        } else {
            HWPrint("发送失败")
        }
    }
    
    
//    func getIFAddresses() -> [String] {
//        var addresses = [String]()
//        // Get list of all interfaces on the local machine:
//        var ifaddr : UnsafeMutablePointer<ifaddrs>? = nil
//        if getifaddrs(&ifaddr) == 0 {
//            var ptr = ifaddr
//            while ptr != nil {
//                let flags = Int32((ptr?.pointee.ifa_flags)!)
//                var addr = ptr?.pointee.ifa_addr.pointee
//                HWPrint(addr)
//                // Check for running IPv4, IPv6 interfaces. Skip the loopback interface.
//                if (flags & (IFF_UP|IFF_RUNNING|IFF_LOOPBACK)) == (IFF_UP|IFF_RUNNING) {
//                    if addr?.sa_family == UInt8(AF_INET) || addr?.sa_family == UInt8(AF_INET6) {
//                        // Convert interface address to a human readable string:
//                        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
//                        if (getnameinfo(&addr!, socklen_t((addr?.sa_len)!), &hostname, socklen_t(hostname.count),
//                                        nil, socklen_t(0), NI_NUMERICHOST) == 0) {
//                            if let address = String(validatingUTF8: hostname) {
//                                addresses.append(address)
//                            }
//                        }
//                    }
//                }
//                ptr = ptr?.pointee.ifa_next
//            }
//            freeifaddrs(ifaddr)
//        }
//        print("Local IP \(addresses)")
//        return addresses
//    }
}

extension GCDSocketServerViewController : GCDAsyncSocketDelegate {
    //接收到请求
    func socket(_ sock: GCDAsyncSocket, didAcceptNewSocket newSocket: GCDAsyncSocket) {
        showLogMsg("收到客服点连接.....")
        let ip = newSocket.connectedHost ?? ""
        let port = newSocket.connectedPort
        let log = String(format: "客户端地址:%@ 端口:%d", ip, port)
        showLogMsg(log)
        self.clientSocket = newSocket
        newSocket.readData(withTimeout: -1, tag: 0)
        if let data = "你连接成功了,大兄弟".data(using: String.Encoding.utf8) {
            self.clientSocket?.write(data, withTimeout: -1, tag: 0)
            showLogMsg("你连接成功了,大兄弟")
        }
    }
    
    func socket(_ sock: GCDAsyncSocket, didRead data: Data, withTag tag: Int) {
        let strMsg = String(bytes: data, encoding: String.Encoding.utf8)
        let log = String(format: "客户端: %@", strMsg ?? "")
        showLogMsg(log)
        sock.readData(withTimeout: -1, tag: 0)
    }
    
}
